### PR TITLE
Fix token counting to use attention mask instead of ids

### PR DESCRIPTION
**Pull request by dakinggg**
This fixes off by one errors when pad token and eos token are the same thing.

Manual test compare to before, there are 1440 more tokens counted for a run of length 30 batches * 48 samples per batch, as expected, because there is one eos token per sample.
Screenshot 2023-12-13 at 4 24 13 PM

### CODE REVIEW

**Review by alextrott16 — APPROVED** [review]
Thanks for doing this. Definitely makes a lot more sense to be counting based on attention mask. Approve.

File: tests/data/test_dataloader.py [hunk 1]
```diff
@@ -735,8 +741,10 @@ def test_token_counting_func_dataloader_setting(
             },
             **common_args
         })
+        ds_mock = MagicMock()
```
Comment (irenedea) [inline tests/data/test_dataloader.py:744]: Why is this mocked but not for streaming-finetuning?

File: tests/data/test_dataloader.py [hunk 2]
```diff
@@ -654,27 +654,33 @@ def test_token_counting_func_dataloader_setting(
...
     if dataloader_type == 'denoising':
-        batch_tokenized['decoder_input_ids'] = batch_tokenized[
-            'input_ids'].clone()
+        expected_token_count += 2 * batch_size  # for the two eos tokens
+        expected_token_count += 5 * batch_size  # for the corruption prefix tokens
+
+    if dataloader_type in {'finetuning-hf', 'finetuning-streaming'}:
+        for b in batch_tokenized:
+            b['labels'] = b['input_ids'].copy()
         expected_token_count *= 2
```
Comment (irenedea) [inline tests/data/test_dataloader.py:682]: why is the expected token count doubled for finetuning?

**Review by irenedea — APPROVED** [review]
LGTM! Just had a couple questions.

File: tests/data/test_dataloader.py (hunk 1 above)
Comment (irenedea) [reply tests/data/test_dataloader.py:744]: oh, nvm it is!

Comment (irenedea) [reply tests/data/test_dataloader.py:744]: I assume we need the mock here because the streamingtextdataset has attributes that are accessed.
